external/llvm-project/mlir/lib/Dialect/LLVMIR/Transforms/SoftwareBf16.cpp

```diff
@@ -0,0 +1,364 @@
+  // isCompatibleType(), which doesn't convert the element type
+  converter.addConversion(
+      [llvmI16, bf16](mlir::VectorType type) -> Optional<Type> {
+        if (type.getElementType() == bf16)
+          return type.clone(llvmI16);
+        return llvm::None; // continue search
+      });
+
+  converter.addConversion(
+      [&](LLVM::LLVMStructType type, SmallVectorImpl<Type> &results,
+          ArrayRef<Type> callStack) -> Optional<LogicalResult> {
+        bool converted = false;
+        SmallVector<Type> convertedElemTypes;
+        convertedElemTypes.reserve(type.getBody().size());
+        for (auto t : type.getBody()) {
+          SmallVector<Type, 1> element;
+          if (failed(converter.convertType(t, element)))
+            return llvm::None;
+          assert(element.size() == 1);
+          convertedElemTypes.push_back(element[0]);
+          if (t != element[0])
+            converted = true;
+        }
+
+        if (!converted) {
+          results.push_back(type);
+          return success();
+        }
+
+        // Identified StructType
+        if (type.isIdentified()) {
+          auto convertedType = LLVM::LLVMStructType::getIdentified(
+              type.getContext(), ("_Converted_" + type.getName()).str());
+          unsigned counter = 1;
+          while (convertedType.isInitialized()) {
+            convertedType = LLVM::LLVMStructType::getIdentified(
+                type.getContext(),
+                ("_Converted_" + Twine(counter++) + type.getName()).str());
+          }
+          if (llvm::count(callStack, type) > 1) {
+            results.push_back(convertedType);
+            return success();
+          }
+          if (failed(
+                  convertedType.setBody(convertedElemTypes, type.isPacked())))
+            return llvm::None;
+          results.push_back(convertedType);
+          return success();
+        }
+
+        // Literal StructType
+        results.push_back(LLVM::LLVMStructType::getLiteral(
+            type.getContext(), convertedElemTypes, type.isPacked()));
+        return success();
+      });
+
+  patterns.add<BF16ConstCasting, SoftwareBF16Trunc, SoftwareBF16Ext>(ctx);
+
+  patterns.add<BF16AsF32<LLVM::FAddOp>, BF16AsF32<LLVM::FCmpOp>,
+               BF16AsF32<LLVM::FDivOp>, BF16AsF32<LLVM::FMulOp>,
+               BF16AsF32<LLVM::FNegOp>, BF16AsF32<LLVM::FRemOp>,
+               BF16AsF32<LLVM::FSubOp>, BF16AsF32<LLVM::FPToSIOp>,
+               BF16AsF32<LLVM::FPToUIOp>, BF16AsF32<LLVM::SIToFPOp>,
+               BF16AsF32<LLVM::UIToFPOp>, BF16AsF32<LLVM::FAbsOp>,
+               BF16AsF32<LLVM::FCeilOp>, BF16AsF32<LLVM::FFloorOp>,
+               BF16AsF32<LLVM::FMAOp>, BF16AsF32<LLVM::FMulAddOp>>(ctx);
+}
+
+namespace {
+struct SoftwareBF16Pass : public SoftwareBF16Base<SoftwareBF16Pass> {
+  void runOnOperation() override {
+    auto m = getOperation();
+    MLIRContext *ctx = m->getContext();
+    LowerToLLVMOptions options(ctx);
```

Comment:
(To state my point differently - I claim that, since we're not converting to LLVM, we don't need of the machinery for that conversion ... at this point, it's already happened, and so the regular `TypeConverter` API will be enough)